BootloaderCorePkg/Library/PciEnumerationLib/PciEnumerationLib.c

```diff
@@ -1136,43 +1144,62 @@ PciEnumeration (
+  UINT16                     PrimaryBus;
+  UINT16                     SubBusNumber;
+  UINT64                     IoBase;
+  UINT64                     Mem32Base;
   PCI_IO_DEVICE             *Root;
 
   SetAllocationPool (MemPool);
 
-  SubBusNumber = 0;
-  Root = CreatePciIoDevice (NULL, NULL, 0, 0, 0);
-  PciScanBus (Root, 0, &SubBusNumber, NULL);
+  IoBase = PcdGet32 (PcdPciResourceIoBase);
+  Mem32Base = PcdGet32 (PcdPciResourceMem32Base);
+  for (PrimaryBus = 0; PrimaryBus <= PCI_MAX_BUS; PrimaryBus = SubBusNumber + 1) {
+    SubBusNumber = PrimaryBus;
+    Root = CreatePciIoDevice (NULL, NULL, PrimaryBus, 0, 0);
+    Status = PciScanBus (Root, PrimaryBus, &SubBusNumber, NULL);
+    if (EFI_ERROR (Status)) {
+      continue;
+    }
 
-  CalculateResource (Root, PciBarTypeIo16);
-  CalculateResource (Root, PciBarTypeMem32);
-  CalculateResource (Root, PciBarTypePMem32);
+    CalculateResource (Root, PciBarTypeIo16);
+    CalculateResource (Root, PciBarTypeMem32);
+    CalculateResource (Root, PciBarTypePMem32);
 
-  for (Idx = 0; Idx < PCI_MAX_BAR; Idx++) {
-    if (Idx == 0) { //IO16
-      Root->PciBar[0].BaseAddress = PcdGet32 (PcdPciResourceIoBase);
-    } else if (Idx == 2) { // MEM32
-      Root->PciBar[2].BaseAddress = PcdGet32 (PcdPciResourceMem32Base);
-    } else if (Idx == 3) { // PMEM32
-      Root->PciBar[3].BaseAddress = ALIGN (Root->PciBar[2].BaseAddress + Root->PciBar[2].Length, Root->PciBar[3].Alignment);
+    for (Idx = 0; Idx < PCI_MAX_BAR; Idx++) {
+      if (Idx == 0) { //IO16
+        Root->PciBar[0].BaseAddress = IoBase;
+      } else if (Idx == 2) { // MEM32
+        Root->PciBar[2].BaseAddress = Mem32Base;
+      } else if (Idx == 3) { // PMEM32
+        Root->PciBar[3].BaseAddress = ALIGN (Root->PciBar[2].BaseAddress + Root->PciBar[2].Length, Root->PciBar[3].Alignment);
+      }
     }
-  }
-  Root->Address = PcdGet32 (PcdPciResourceMem32Base) + SubBusNumber;
 
-  ProgramResource (Root, PciBarTypeIo16);
-  ProgramResource (Root, PciBarTypeMem32);
-  ProgramResource (Root, PciBarTypePMem32);
+    Root->Address = PcdGet32 (PcdPciResourceMem32Base) + (PrimaryBus << 8) + SubBusNumber;
 
-  EnablePciDevice (Root);
+    ProgramResource (Root, PciBarTypeIo16);
+    ProgramResource (Root, PciBarTypeMem32);
+    ProgramResource (Root, PciBarTypePMem32);
+
+    EnablePciDevice (Root);
 
 #if DEBUG_PCI_ENUM
-  DumpDeivceTree  (Root);
-  DumpResourceBar (Root);
-  DEBUG ((DEBUG_INFO, "MEM Pool Used: 0x%08X\n", (UINT32)GetAllocationPool() - (UINT32)MemPool));
+    DumpDeivceTree  (Root);
+    DumpResourceBar (Root);
+    DEBUG ((DEBUG_INFO, "MEM Pool Used: 0x%08X\n", (UINT32)GetAllocationPool() - (UINT32)MemPool));
 #endif
 
+    //
+    // Io/Mem32 Base for next bus
+    //
+    IoBase = ALIGN (Root->PciBar[0].BaseAddress + Root->PciBar[0].Length, Root->PciBar[0].Alignment);
+    Mem32Base = MAX (Mem32Base, ALIGN (Root->PciBar[2].BaseAddress + Root->PciBar[2].Length, Root->PciBar[2].Alignment));

```

Comment:
MEM and PMEM should not be allocated in interleaved manner.  All MEM needs to be a single range.  And all PMEM needs to be a different range.   It is to allow WB cache type for preferable memory.

Reply:
Thanks for the feedback. I am going to re-visit this PR.